### PR TITLE
SoMe issuer

### PR DESCRIPTION
**Pull request by lassemoldrup**
## Purpose

Addresses #16.

## Changes

Two issuers of Web3 ID credentials are created, one for Telegram and one for Discord.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.

### CODE REVIEW

File: examples/some-issuer/json-schemas/discord/JsonSchema2023-discord.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+          "type": "string",
+          "description": "Discord user id",
+          "index": "0"
+        }
+      },
+      "required": ["id", "userId"]
+    }
+  },
+  "required": ["credentialSubject"]
+}
```
[the remark this answers is not in the document]
Comment (annenkov) [reply examples/some-issuer/json-schemas/discord/JsonSchema2023-discord.json:27]: The `index` is no longer needed, the attributes are now identified by their "string" names.

**Review by annenkov — APPROVED** [review]
Looks good.
Just added some comments about the JSON schema examples, I don't think this affects the functionality though.

File: examples/some-issuer/json-schemas/discord/credential-metadata.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "title": "Discord Certificate",
+  "logo": {
+    "url": "https://avatars.githubusercontent.com/u/39614219?s=200&v=4"
+  },
+  "background_color": "#7289da"
```
Comment (soerenbf) [inline examples/some-issuer/json-schemas/discord/credential-metadata.json:6]: Remember to change this

```suggestion
  "backgroundColor": "#7289da"
```

File: examples/some-issuer/src/bin/discord.rs [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+struct AccessTokenRequestData<'a> {
+    client_id: &'a str,
+    client_secret: &'a str,
+    grant_type: &'static str,
+    code: &'a str,
+    redirect_uri: &'a str,
+}
+
+#[derive(Deserialize)]
+#[allow(unused)]
```
Comment (abizjak) [inline examples/some-issuer/src/bin/discord.rs:157]: Why is this necessary?

File: examples/some-issuer/src/bin/discord.rs [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+    State(state): State<AppState>,
+    Query(params): Query<HashMap<String, String>>,
+    session: WritableSession,
+) -> Result<Html<String>, StatusCode> {
+    async fn respond(
+        state: AppState,
+        params: HashMap<String, String>,
+        mut session: WritableSession,
+    ) -> anyhow::Result<Html<String>> {
+        let code = params.get("code").context("Missing 'code' query param")?;
```
Comment (abizjak) [inline examples/some-issuer/src/bin/discord.rs:183]: Why is this not handled directly in the type? So instead of
```
Query(HashMap<...>)
```
you'd take a
```
Query(Oauth2RedirectParams)
```
where

```
#[derive(serde::Deserialize)]
struct Oauth2RedirectParams {
   code: String
}
```

File: examples/some-issuer/src/bin/telegram.rs [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    auth_date: u64,
+    hash: String,
+}
+
+type HmacSha256 = Hmac<Sha256>;
+impl User {
+    fn check(&self, telegram_bot_token: &str) -> anyhow::Result<()> {
+        let mut hasher = Sha256::new();
+        hasher.update(telegram_bot_token);
+        let key = hasher.finalize();
```
Comment (abizjak) [inline examples/some-issuer/src/bin/telegram.rs:137]:
```
let key = Sha256::digest(telegram_bot_token.as_bytes());
```

File: examples/some-issuer/src/bin/telegram.rs [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+        let mut mac = HmacSha256::new_from_slice(&key).unwrap();
```
Comment (abizjak) [inline examples/some-issuer/src/bin/telegram.rs:139]: Document that we know the length is correct so this will never fail.

File: examples/some-issuer/src/bin/telegram.rs [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        let expected = hex::decode(&self.hash)?;
+
+        anyhow::ensure!(&mac_bytes[..] == &expected[..], "MAC did not match hash.");
+        Ok(())
+    }
+}
+
+/// Note: the exact format of this is important, as it is used for verification.
+/// See https://core.telegram.org/widgets/login
+impl Display for User {
```
Comment (abizjak) [inline examples/some-issuer/src/bin/telegram.rs:152]: It seems odd to use a `Display` implementation for such a sensitive thing.

It would be better to use a dedicated function. Display usually does not have such strict requirements.

File: examples/some-issuer/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    tracing::info!("Request to issue a credential.");
+
+    if let Err(err) = validate_credential(&issuer, &credential) {
+        tracing::warn!("Failed to validate credential: {err}");
+        return Err(StatusCode::BAD_REQUEST);
+    }
+
+    match register_credential(issuer, credential, user_id).await {
+        Ok(res) => {
+            tracing::info!("Successfully issued credential.");
```
Comment (abizjak) [inline examples/some-issuer/src/lib.rs:56]: Some more details about the credential should be here.

And this should be at debug I think.

File: examples/some-issuer/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+/// Checks that the credential is reasonable.
+fn validate_credential(issuer: &IssuerState, credential: &CredentialInfo) -> anyhow::Result<()> {
+    anyhow::ensure!(
+        credential.holder_revocable,
+        "Credential should be holder revocable."
+    );
+    let now = chrono::Utc::now().timestamp_millis();
+    let delta =
+        Timestamp::from_timestamp_millis(now as u64).duration_between(credential.valid_from);
```
Comment (abizjak) [inline examples/some-issuer/src/lib.rs:74]:
```suggestion
    let now = chrono::Utc::now().timestamp_millis();
    let delta = now.signed_duration_since(credential.valid_from);
```

File: .gitmodules [hunk 10]
```diff
@@ -1,3 +1,7 @@
 [submodule "deps/concordium-rust-sdk"]
 	path = deps/concordium-rust-sdk
 	url = ../concordium-rust-sdk.git
+[submodule "deps/concordium-browser-wallet"]
+	path = deps/concordium-browser-wallet
+	url = ../concordium-browser-wallet.git
+	branch = web3Id-all-together
```
Comment (abizjak) [inline .gitmodules:7]: There is a feature branch now that should be used. `feature/web3-id` 

That's where everything is at.

**Review by abizjak — COMMENTED** [review]
Generally looks good.

Before continuing work on this we should do the following
- update Rust SDK to the latest main
- get the test build of the new browser wallet that supports the new formats
- update all the JSON schemas etc to those formats

I did not go through the entire flow with a security aspect in mind now, so we'll want to make sure that that is all good. But for now that is not the most important aspect.

One general aspect that needs work is documentation, but since we are stressed for time we'll do that at the end :)

File: examples/some-issuer/src/lib.rs (hunk 8 above)
Comment (lassemoldrup) [reply examples/some-issuer/src/lib.rs:56]: In the version I am working on it prints the user id and username, is that the kind of thing you were referring to?

> And this should be at debug I think.

I assume the corresponding "Request to issue a credential." and "" Using nonce... to send transaction." also should be debug then.

File: examples/some-issuer/src/lib.rs (hunk 9 above)
Comment (lassemoldrup) [reply examples/some-issuer/src/lib.rs:74]: The suggestion doesn't work (I assume you also meant to remove `.timestamp_millis()`). `credential.valid_from` is of type `concordium_contracts_common::Timestamp`.

File: examples/some-issuer/src/lib.rs (hunk 8 above)
Comment (abizjak) [reply examples/some-issuer/src/lib.rs:56]: Yes.

**Comment by soerenbf** [comment]
Comments addressed in https://github.com/Concordium/concordium-web3id/pull/68
